src/shared_types.rs

```diff
@@ -0,0 +1,224 @@
+        }
+    }
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum DataAddress {
+    Map(Address),
+    Sequence(Address),
+}
+
+/// The address of a data type.
+/// Each scope has its own address space.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Address {
+    /// Address in the public namespace.
+    Public {
+        /// Name.
+        name: XorName,
+        /// Type tag.
+        tag: u64,
+    },
+    /// Address in the private namespace.
+    Private {
+        /// Name.
+        name: XorName,
+        /// Type tag.
+        tag: u64,
+    },
+}
+
+impl Address {
+    /// Returns an Address instance for the specified data type scope.
+    pub fn from_scope(scope: Scope, name: XorName, tag: u64) -> Self {
+        match scope {
+            Scope::Public => Address::Public { name, tag },
+            Scope::Private => Address::Private { name, tag },
+        }
+    }
+
+    /// Returns the scope of data type that this address space represents.
+    pub fn scope(&self) -> Scope {
+        match self {
+            Address::Public { .. } => Scope::Public,
+            Address::Private { .. } => Scope::Private,
+        }
+    }
+
+    /// Returns the XorName that the data is located by
+    /// within its address space.
+    pub fn name(&self) -> &XorName {
+        match self {
+            Address::Public { ref name, .. } | Address::Private { ref name, .. } => name,
+        }
+    }
+
+    /// Returns the tag that denotes a specific address space.
+    pub fn tag(&self) -> u64 {
+        match self {
+            Address::Public { tag, .. } | Address::Private { tag, .. } => *tag,
+        }
+    }
+
+    /// Returns true if the address is for a public data type.
+    pub fn is_public(&self) -> bool {
+        self.scope().is_public()
+    }
+
+    /// Returns true if the address is for a private data type.
+    pub fn is_private(&self) -> bool {
+        self.scope().is_private()
+    }
+
+    /// Returns the Address serialised and encoded in z-base-32.
+    pub fn encode_to_zbase32(&self) -> String {
+        utils::encode(&self)
+    }
+
+    /// Creates from z-base-32 encoded string.
+    pub fn decode_from_zbase32<I: Decodable>(encoded: I) -> Result<Self> {
+        utils::decode(encoded)
+    }
+}
+
+pub fn to_absolute_version(version: Version, count: usize) -> Option<usize> {
+    match version {
+        Version::FromStart(version) if (version as usize) <= count => Some(version as usize),
+        Version::FromStart(_) => None,
+        Version::FromEnd(version) => count.checked_sub(version as usize),
+    }
+}
+
+pub fn to_absolute_range(start: Version, end: Version, count: usize) -> Option<Range<usize>> {
+    let start = to_absolute_version(start, count)?;
+    let end = to_absolute_version(end, count)?;
+
+    if start <= end {
+        Some(start..end) // `end` is exclusive in Range
```

Comment:
```suggestion
        Some(start..end)
```
Since `+1` is removed, the comment is not required.

Reply:
fixed